src/web/releases.rs

```diff
@@ -561,11 +561,25 @@ pub fn search_handler(req: &mut Request) -> IronResult<Response> {
             return redirect_to_random_crate(req, &mut conn);
         }
 
-        let (krate, query) = match query.split_once("::") {
+        let (krate, mut query) = match query.split_once("::") {
             Some((krate, query)) => (krate.to_string(), format!("?search={query}")),
             None => (query.clone(), "".to_string()),
         };
 
+        for (k, v) in params
```

Comment:
this really looks like we should switch `query` to be a list of pairs, hashmap or similar, 

and use [`Url::parse_with_params`](https://docs.rs/iron/latest/iron/url/struct.Url.html#examples-1) below instead of this manual query string construction

Reply:
If I do this, the query params gets url encoded,

so `?search=crate::path` becomes `?search=crate%3A%3Apath`

not sure this is avoidable

here's the diff from d333ab50302686da8a7c179c754cbbc0c7e0aca0

(without fixed tests)

```diff
diff --git a/src/web/releases.rs b/src/web/releases.rs
index 72c13fe8..148f8e3a 100644
--- a/src/web/releases.rs
+++ b/src/web/releases.rs
@@ -561,24 +561,22 @@ pub fn search_handler(req: &mut Request) -> IronResult<Response> {
             return redirect_to_random_crate(req, &mut conn);
         }
 
-        let (krate, mut query) = match query.split_once("::") {
-            Some((krate, query)) => (krate.to_string(), format!("?search={query}")),
-            None => (query.clone(), "".to_string()),
-        };
+        let mut queries = HashMap::new();
 
-        for (k, v) in params
-            .iter()
-            .filter(|(k, _)| !matches!(k.as_ref(), "i-am-feeling-lucky" | "query"))
-        {
-            if query.is_empty() {
-                query.push('?');
-            } else {
-                query.push('&')
+        let krate = match query.split_once("::") {
+            Some((krate, query)) => {
+                queries.insert("search", query);
+                krate.to_string()
             }
-            query.push_str(k);
-            query.push('=');
-            query.push_str(v);
-        }
+            None => query.clone(),
+        };
+
+        queries.extend(
+            params
+                .iter()
+                .filter(|(k, _)| !matches!(k.as_ref(), "i-am-feeling-lucky" | "query"))
+                .map(|(k, v)| (k.as_ref(), v.as_ref())),
+        );
 
         // since we never pass a version into `match_version` here, we'll never get
         // `MatchVersion::Exact`, so the distinction between `Exact` and `Semver` doesn't
@@ -590,10 +588,18 @@ pub fn search_handler(req: &mut Request) -> IronResult<Response> {
             let base = redirect_base(req);
             let url = if matchver.rustdoc_status {
                 let target_name = matchver.target_name;
-                ctry!(
-                    req,
-                    Url::parse(&format!("{base}/{krate}/{version}/{target_name}/{query}"))
-                )
+                let path = format!("{base}/{krate}/{version}/{target_name}/");
+                if queries.is_empty() {
+                    ctry!(req, Url::parse(&path))
+                } else {
+                    ctry!(
+                        req,
+                        Url::from_generic_url(ctry!(
+                            req,
+                            iron::url::Url::parse_with_params(&path, queries)
+                        ))
+                    )
+                }
             } else {
                 ctry!(req, Url::parse(&format!("{base}/crate/{krate}/{version}")))
             };
diff --git a/src/web/rustdoc.rs b/src/web/rustdoc.rs
index a99219df..9e5993c8 100644
--- a/src/web/rustdoc.rs
+++ b/src/web/rustdoc.rs
@@ -42,30 +42,29 @@ static DOC_RUST_LANG_ORG_REDIRECTS: Lazy<HashMap<&str, &str>> = Lazy::new(|| {
 pub fn rustdoc_redirector_handler(req: &mut Request) -> IronResult<Response> {
     fn redirect_to_doc(
         req: &Request,
-        mut url_str: String,
+        url_str: String,
         permanent: bool,
         path_in_crate: Option<&str>,
     ) -> IronResult<Response> {
-        let mut question_mark = false;
+        let mut queries: HashMap<std::borrow::Cow<'_, str>, std::borrow::Cow<'_, str>> =
+            HashMap::new();
         if let Some(path) = path_in_crate {
-            url_str.push_str("?search=");
-            url_str.push_str(path);
-            question_mark = true;
+            queries.insert("search".into(), path.into());
         }
-        if let Some(query) = req.url.query() {
-            if !question_mark {
-                url_str.push('?');
-            } else {
-                url_str.push('&');
-            }
-            url_str.push_str(query);
-        }
-        let url = ctry!(req, Url::parse(&url_str));
+        let url: iron::url::Url = req.url.clone().into();
+        let query_pairs = url.query_pairs();
+        queries.extend(query_pairs);
+        let url = if queries.is_empty() {
+            ctry!(req, iron::url::Url::parse(&url_str))
+        } else {
+            ctry!(req, iron::url::Url::parse_with_params(&url_str, queries))
+        };
         let (status_code, max_age) = if permanent {
             (status::MovedPermanently, 86400)
         } else {
             (status::Found, 0)
         };
+        let url = ctry!(req, Url::from_generic_url(url));
         let mut resp = Response::with((status_code, Redirect(url)));
         resp.headers
             .set(CacheControl(vec![CacheDirective::MaxAge(max_age)]));
```



<details>
<summary>this fails with</summary>

```
failures:

---- web::releases::tests::search_coloncolon_path_redirects_to_crate_docs stdout ----
[2022-09-17T17:57:39.346Z ERROR docs_rs::web::page::templates] Failed to load rustc resource suffix: failed to parse the rustc version
[2022-09-17T17:57:39.346Z WARN  docs_rs::repositories::github] did not collect `github.com` stats as no token was provided
[2022-09-17T17:57:39.346Z WARN  docs_rs::repositories::gitlab] will try to retrieve `gitlab.com` stats without token since none was provided
[2022-09-17T17:57:39.346Z WARN  docs_rs::repositories::gitlab] will try to retrieve `gitlab.freedesktop.org` stats without token since none was provided
[2022-09-17T17:57:39.348Z INFO  docs_rs::web] Running docs.rs web server on http://127.0.0.1:59062
[2022-09-17T17:57:39.348Z DEBUG docs_rs::test::fakes] before upload source
[2022-09-17T17:57:39.349Z DEBUG docs_rs::test::fakes] adding directory Sources from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeQfh5xX
[2022-09-17T17:57:39.367Z DEBUG docs_rs::test::fakes] added source files []
[2022-09-17T17:57:39.367Z DEBUG docs_rs::test::fakes] writing file /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-faketp4yd5/some_random_crate/index.html
[2022-09-17T17:57:39.367Z DEBUG docs_rs::test::fakes] added rustdoc files
[2022-09-17T17:57:39.367Z DEBUG docs_rs::test::fakes] adding directory Rustdoc from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-faketp4yd5
[2022-09-17T17:57:39.403Z DEBUG docs_rs::test::fakes] uploaded rustdoc files: [["text/html","some_random_crate/index.html"]]
[2022-09-17T17:57:39.406Z DEBUG docs_rs::db::add_package] Adding package into database
[2022-09-17T17:57:39.689Z INFO  docs_rs::db::add_package] Updating crate data for some_random_crate
[2022-09-17T17:57:39.726Z DEBUG docs_rs::db::add_package] Adding build into database
[2022-09-17T17:57:40.074Z DEBUG docs_rs::test::fakes] before upload source
[2022-09-17T17:57:40.074Z DEBUG docs_rs::test::fakes] adding directory Sources from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakebIe4uJ
[2022-09-17T17:57:40.083Z DEBUG docs_rs::test::fakes] added source files []
[2022-09-17T17:57:40.083Z DEBUG docs_rs::test::fakes] writing file /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeqpCw5a/some_other_crate/index.html
[2022-09-17T17:57:40.083Z DEBUG docs_rs::test::fakes] added rustdoc files
[2022-09-17T17:57:40.083Z DEBUG docs_rs::test::fakes] adding directory Rustdoc from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeqpCw5a
[2022-09-17T17:57:40.097Z DEBUG docs_rs::test::fakes] uploaded rustdoc files: [["text/html","some_other_crate/index.html"]]
[2022-09-17T17:57:40.100Z DEBUG docs_rs::db::add_package] Adding package into database
[2022-09-17T17:57:40.171Z INFO  docs_rs::db::add_package] Updating crate data for some_other_crate
[2022-09-17T17:57:40.205Z DEBUG docs_rs::db::add_package] Adding build into database
[2022-09-17T17:57:40.226Z DEBUG docs_rs::test] getting http://127.0.0.1:59062/releases/search?query=some_random_crate::somepath (no redirects)
[2022-09-17T17:57:40.243Z DEBUG docs_rs::test] getting http://127.0.0.1:59062/some_random_crate/1.0.0/some_random_crate/?search=somepath (no redirects)
[2022-09-17T17:57:40.253Z DEBUG docs_rs::web::metrics] rendering time - match version: 7.576916ms
[2022-09-17T17:57:40.287Z DEBUG docs_rs::web::metrics] rendering time - crate details: 34.036416ms
[2022-09-17T17:57:40.331Z DEBUG docs_rs::web::metrics] rendering time - fetch from storage: 44.065458ms
[2022-09-17T17:57:40.331Z DEBUG docs_rs::web::metrics] rendering time - find latest path: 77.208µs
[2022-09-17T17:57:40.333Z DEBUG docs_rs::web::metrics] rendering time - rewrite html: 1.835375ms
[2022-09-17T17:57:40.334Z DEBUG docs_rs::test] getting http://127.0.0.1:59062/releases/search?query=some_random_crate::some::path (no redirects)
the test failed: expected redirect from /releases/search?query=some_random_crate::some::path to /some_random_crate/1.0.0/some_random_crate/?search=some::path
  caused by: expected redirect from /releases/search?query=some_random_crate::some::path to /some_random_crate/1.0.0/some_random_crate/?search=some::path
  caused by: got redirect to /some_random_crate/1.0.0/some_random_crate/?search=some%3A%3Apath
disabled backtrace
thread 'web::releases::tests::search_coloncolon_path_redirects_to_crate_docs' panicked at 'the test failed', src/test/mod.rs:39:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- web::rustdoc::test::test_redirect_crate_coloncolon_path stdout ----
[2022-09-17T17:58:13.424Z ERROR docs_rs::web::page::templates] Failed to load rustc resource suffix: failed to parse the rustc version
[2022-09-17T17:58:13.424Z WARN  docs_rs::repositories::github] did not collect `github.com` stats as no token was provided
[2022-09-17T17:58:13.424Z WARN  docs_rs::repositories::gitlab] will try to retrieve `gitlab.com` stats without token since none was provided
[2022-09-17T17:58:13.424Z WARN  docs_rs::repositories::gitlab] will try to retrieve `gitlab.freedesktop.org` stats without token since none was provided
[2022-09-17T17:58:13.426Z INFO  docs_rs::web] Running docs.rs web server on http://127.0.0.1:59824
[2022-09-17T17:58:13.426Z DEBUG docs_rs::test::fakes] before upload source
[2022-09-17T17:58:13.426Z DEBUG docs_rs::test::fakes] adding directory Sources from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fake2YMlhq
[2022-09-17T17:58:13.435Z DEBUG docs_rs::test::fakes] added source files []
[2022-09-17T17:58:13.435Z DEBUG docs_rs::test::fakes] writing file /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeC3G4vm/some_random_crate/index.html
[2022-09-17T17:58:13.435Z DEBUG docs_rs::test::fakes] added rustdoc files
[2022-09-17T17:58:13.435Z DEBUG docs_rs::test::fakes] adding directory Rustdoc from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeC3G4vm
[2022-09-17T17:58:13.480Z DEBUG docs_rs::test::fakes] uploaded rustdoc files: [["text/html","some_random_crate/index.html"]]
[2022-09-17T17:58:13.482Z DEBUG docs_rs::db::add_package] Adding package into database
[2022-09-17T17:58:13.792Z INFO  docs_rs::db::add_package] Updating crate data for some_random_crate
[2022-09-17T17:58:13.837Z DEBUG docs_rs::db::add_package] Adding build into database
[2022-09-17T17:58:14.100Z DEBUG docs_rs::test::fakes] before upload source
[2022-09-17T17:58:14.100Z DEBUG docs_rs::test::fakes] adding directory Sources from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakebQCXlZ
[2022-09-17T17:58:14.106Z DEBUG docs_rs::test::fakes] added source files []
[2022-09-17T17:58:14.106Z DEBUG docs_rs::test::fakes] writing file /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeFhiscs/some_other_crate/index.html
[2022-09-17T17:58:14.106Z DEBUG docs_rs::test::fakes] added rustdoc files
[2022-09-17T17:58:14.106Z DEBUG docs_rs::test::fakes] adding directory Rustdoc from /var/folders/l6/fdwj1xb53mlcfq5jk_z_6yp80000gn/T/docs.rs-fakeFhiscs
[2022-09-17T17:58:14.121Z DEBUG docs_rs::test::fakes] uploaded rustdoc files: [["text/html","some_other_crate/index.html"]]
[2022-09-17T17:58:14.123Z DEBUG docs_rs::db::add_package] Adding package into database
[2022-09-17T17:58:14.266Z INFO  docs_rs::db::add_package] Updating crate data for some_other_crate
[2022-09-17T17:58:14.295Z DEBUG docs_rs::db::add_package] Adding build into database
[2022-09-17T17:58:14.321Z DEBUG docs_rs::test] getting http://127.0.0.1:59824/some_random_crate::somepath (no redirects)
[2022-09-17T17:58:14.334Z DEBUG docs_rs::web::metrics] rendering time - match version: 9.985583ms
[2022-09-17T17:58:14.338Z DEBUG docs_rs::web::metrics] rendering time - fetch release doc status: 3.893958ms
[2022-09-17T17:58:14.338Z DEBUG docs_rs::web::metrics] rendering time - redirect to doc: 134.541µs
[2022-09-17T17:58:14.338Z DEBUG docs_rs::test] getting http://127.0.0.1:59824/some_random_crate/latest/some_random_crate/?search=somepath (no redirects)
[2022-09-17T17:58:14.349Z DEBUG docs_rs::web::metrics] rendering time - match version: 8.385416ms
[2022-09-17T17:58:14.371Z DEBUG docs_rs::web::metrics] rendering time - crate details: 21.974958ms
[2022-09-17T17:58:14.425Z DEBUG docs_rs::web::metrics] rendering time - fetch from storage: 53.414666ms
[2022-09-17T17:58:14.425Z DEBUG docs_rs::web::metrics] rendering time - find latest path: 81.291µs
[2022-09-17T17:58:14.427Z DEBUG docs_rs::web::metrics] rendering time - rewrite html: 1.790833ms
[2022-09-17T17:58:14.428Z DEBUG docs_rs::test] getting http://127.0.0.1:59824/some_random_crate::some::path (no redirects)
[2022-09-17T17:58:14.441Z DEBUG docs_rs::web::metrics] rendering time - match version: 9.479416ms
[2022-09-17T17:58:14.445Z DEBUG docs_rs::web::metrics] rendering time - fetch release doc status: 3.736125ms
[2022-09-17T17:58:14.445Z DEBUG docs_rs::web::metrics] rendering time - redirect to doc: 159.291µs
the test failed: expected redirect from /some_random_crate::some::path to /some_random_crate/latest/some_random_crate/?search=some::path
  caused by: expected redirect from /some_random_crate::some::path to /some_random_crate/latest/some_random_crate/?search=some::path
  caused by: got redirect to /some_random_crate/latest/some_random_crate/?search=some%3A%3Apath
disabled backtrace
thread 'web::rustdoc::test::test_redirect_crate_coloncolon_path' panicked at 'the test failed', src/test/mod.rs:39:9


failures:
    web::releases::tests::search_coloncolon_path_redirects_to_crate_docs
    web::rustdoc::test::test_redirect_crate_coloncolon_path
```

</details>

Reply:
When I test this with a real doc-page: 

https://docs.rs/itertools/latest/itertools/?search=some%3A%3Apath

it looks like the encoded search will be correctly decoded, which means the test assertion should be adapted. 

A manual test would be good anyways :)

Reply:
So that behaviour is ok? I'm mostly thinking about it in a UX experience, i personally prefer it if it was not encoded.